transcom/mymove
# INT B-20898 sit entry date & FADD validation on update and creation

## [Agility ticket](https://www13.v1host.com/USTRANSCOM38/assetdetail.v1?number=B-20898)

## Summary

When the Prime requests SIT for destination, we want to make sure that the SIT entry date is not before the provided First Available Delivery Date.
In addition to creation, we also need to consider this logic when SIT service items are updated (resubmitted).

This PR does the following:
- adds `UnprocessableEntity` return with error to the prime `mto_service_item` `CreateMTOServiceItemHandler` handler
- added new section in `mto_service_item_validators.go` and placed existing relevant validators in that section
- added validator for creation of DDFSIT (destination SIT) & SIT dest service item updates that handles logic of ensuring the SIT entry date is on or after any of the FADDs (max of 2)
- updates and adds all relevant tests

### How to test

1. Fire up MilMove as an office user
2. Either create or find a move that contains an HHG shipment
3. Go into the Prime simulator and click "Add service item"
4. Add destination SIT and fill out all the data, but make sure the SIT entry date is BEFORE any of the First/second available delivery dates
5. Try to submit - you should get an error back
6. Now change the SIT entry date to be AFTER the FADDs and submit - it should succeed
7. Now go in as the TOO and reject all of those service items
8. You'll need postman for this next part to test this validation on updating
9. Using the endpoint `http://primelocal:3000/prime/v1/mto-service-items/{mtoServiceItemID}`
10. Find the `id`, `eTag`, `reServiceCode` and `sitEntryDate` values of the service item you want to update
11. In your headers, ensure you have `X-CSRF-Token` with the `masked_gorilla` cookie value
12. Also in your headers, ensure you have `If-Match` with the `eTag` value
13. Now in your body, you want to send something like this for each dest SIT service item:
```
{
    "updateReason": "changing SIT entry date",
    "requestApprovalsRequestedStatus": true,
    "modelType": "UpdateMTOServiceItemSIT",
    "sitEntryDate": "2024-08-11",
    "reServiceCode": "DDASIT"
}
```
14. The goal here is to try and resubmit these by changing the SIT entry date to be before the FADD and make sure you get an error
15. Then try again with a SIT entry date that's after the FADDs and ensure it is successful
16. Rinse and repeat for each destination SIT service item

## Screenshots
![Screenshot 2024-08-09 at 2 13 14 PM](https://github.com/user-attachments/assets/01fc8bf2-098e-4e61-a54b-54c09548911f)
